### PR TITLE
Added a note to the documentation to caution against mutating arrays while iterating

### PR DESCRIPTION
**Pull request by MattAlp**
This is unsupported, but not disallowed. Due to its modification of the underlying array, the "next" element yielded to the block by a call to `each` can be confusing. cc @eregon

### CODE REVIEW

**Comment by MattAlp** [comment]
cc @zverok for review

File: array.c [hunk 1]
```diff
@@ -2493,6 +2493,7 @@ ary_enum_length(VALUE ary, VALUE args, VALUE eobj)
  *  Allows the array to be modified during iteration:
  *    a = [:foo, 'bar', 2]
  *    a.each {|element| puts element; a.clear if element.to_s.start_with?('b') }
```
Comment (eregon) [inline array.c:2495]: I would remove these 3 lines, AFAIK they were never explicitly supported (and same below).
The documentation should not show an example of something that should be avoided.

Comment (eregon) [reply array.c:2495]: BTW that example was added by in #3216
@BurdetteLamar FYI, mutating while iterating is explicitly discouraged by matz: http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/23633

Comment (BurdetteLamar) [reply array.c:2495]: > I would remove these 3 lines, AFAIK they were never explicitly supported (and same below). The documentation should not show an example of something that should be avoided.

I'm fine with removing all these.

**Comment by eregon** [comment]
So maybe we should add an example below the new sentence which shows it can go into an infinite loop or yield an element twice, or stop early?
I'm not sure how much "counter examples" make sense, especially for such a simple method.

> things that are discouraged by one team might be the favorite idiom for another.

I think the official stance of *the Ruby core team* and [of matz](http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/23633) about this is mutation during iterating is not officially supported, and I think that opinion makes sense to be reflected in the docs.
And part of the idea behind that is the semantics could change, and e.g. an Array#each infinite loop might no longer be infinite in the future, etc. But I agree for Array#each it sounds unlikely to change at this point, and the current semantics are the simplest to implement for CRuby so it will probably stick to that for a while.
IIRC iteration with some other core methods (maybe on Hash?) does trigger an exception or hard-to-explain results, but it's been a while I ran into that and I don't remember the details.

I think mutating an array while in Array#each is basically a design flaw and will only lead to bugs.
Arguing each_with_index would be more complex, because e.g., it would be fine to assign existing indices (and useful), but for each I don't see any reasonable use case.

**Comment by zverok** [comment]
@eregon To clarify what I was trying to say (Because I've mixed generic considerations about docs, and my stance on this particular method; that's because I am only starting to be systematically involved with docs, so I am a bit more chatty than necessary for a particular case, to align with others)

1. I don't think "changing an array while iterating" is a good practice, and I don't imply I expect that some team has it as a regular trick in their codebase :grinning: 
2. **If** it is the current consensus of the core team that iterating and simultaneously changing behavior is not guaranteed (I know there was a lot of discussion in the past, I was subscribed to ruby-core for I don't remember how long; but that's Matz's post, say, is from twelve years ago, right?)—So, if it is the consensus, I think it is totally OK to state it this way: "So and so, don't do this because _the result is not guaranteed_ even if it worked for you today". 
3. ...but if the consensus has shifted to "OK, we are officially supporting it now" (I'll look into the history of docs/code tomorrow about that), the wording should be different.
4. With that said, I still maintain that—for any of the features—we should try to avoid recommending/discouraging something, just objectively describe the advantages and caveats of some approaches and behaviors, right?..

**Comment by MattAlp** [comment]
Closing due to inactivity

**Pull request closed without merging**